File: 20-valid-parentheses/solution.py
import unittest


class Solution:
    """有效的括号"""

    def isValid(self, s: str) -> bool:
        m = {")": "(", "]": "[", "}": "{"}
        stack = []
        for c in s:
            if c not in m:  # 左括号直接入栈
                stack.append(c)
            elif not stack or m[c] != stack.pop():  # 右括号需要判断
                return False
        return not stack


class TestIsValid(unittest.TestCase):
    def test_isValid(self):
        self.assertEqual(Solution().isValid("()"), True)
        self.assertEqual(Solution().isValid("()[]{}"), True)
        self.assertEqual(Solution().isValid("(]"), False)
        self.assertEqual(Solution().isValid("([)]"), False)
        self.assertEqual(Solution().isValid("{[]}"), True)
        self.assertEqual(Solution().isValid("]"), False)


if __name__ == "__main__":
    unittest.main()

File: 20-valid-parentheses/solution.go
package solution

func isValid(s string) bool {
	stack := make([]rune, 0)
	m := map[rune]rune{
		')': '(',
		']': '[',
		'}': '{',
	}
	for _, c := range s {
		left, ok := m[c]
		if !ok {
			stack = append(stack, c)
		} else if len(stack) == 0 || stack[len(stack)-1] != left {
			return false
		} else {
			stack = stack[:len(stack)-1]
		}
	}
	return len(stack) == 0
}
